### ISSUE TITLE
Unable to lex mysql queries with escaped single quote

### ISSUE THREAD

**Issue by jblotus** [sqlfluff/sqlfluff#3421]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The following sql is unable to be lexed when I attempted to do so:

```sql
UPDATE foo SET bar = 'baz\'s';
```
The console output gave warning:
```shell
WARNING    Zero length Lex item returned from '<unlexable>'. Report this as a bug.
L:   0 | P:   0 |  LXR | Fatal. Unable to lex characters: 0
```

### Expected Behaviour

I would like this to be lexed.

### Observed Behaviour

As a workaround we could use double quotes and avoid escaping which does work properly.
We can probably use -- noqa as well.
```sql
UPDATE foo SET bar = "baz's";
```
However this trigger some other rules

```sql
UPDATE foo SET bar = "baz's and \"bars\" ";
```

```shell
L:   3 | P:  22 | L064 | Inconsistent use of preferred quote style 'double
                       | quotes'. Use 'baz\'s and "bars" ' instead of "baz's and
                       | \"bars\" ".
```

### How to reproduce

The queries are supplied in the description.

### Dialect

Mysql

### Version

sqlfluff, version 0.13.2

### Configuration

default configuration

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by mdahlman** [comment]
@jblotus mentioned that this is a `mysql` issue. He's right. To be more explicit, it's NOT an issue in `ansi`:
```
# Error in mysql
echo "UPDATE foo SET bar = 'baz\'s';" | sqlfluff parse --dialect mysql -
WARNING    Zero length Lex item returned from '<unlexable>'. Report this as a bug.
...Failed to Parse...
==== parsing violations ====
L:   0 | P:   0 |  LXR | Fatal. Unable to lex characters: 0

# happiness in ansi
echo "UPDATE foo SET bar = 'baz\'s';" | sqlfluff parse --dialect ansi -
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        update_statement:
[L:  1, P:  1]      |            keyword:                                          'UPDATE'
[L:  1, P:  7]      |            whitespace:                                       ' '
[L:  1, P:  8]      |            table_reference:
[L:  1, P:  8]      |                identifier:                                   'foo'
[L:  1, P: 11]      |            whitespace:                                       ' '
[L:  1, P: 12]      |            set_clause_list:
[L:  1, P: 12]      |                keyword:                                      'SET'
[L:  1, P: 15]      |                [META] indent:
[L:  1, P: 15]      |                whitespace:                                   ' '
[L:  1, P: 16]      |                set_clause:
[L:  1, P: 16]      |                    column_reference:
[L:  1, P: 16]      |                        identifier:                           'bar'
[L:  1, P: 19]      |                    whitespace:                               ' '
[L:  1, P: 20]      |                    comparison_operator:
[L:  1, P: 20]      |                        raw_comparison_operator:              '='
[L:  1, P: 21]      |                    whitespace:                               ' '
[L:  1, P: 22]      |                    literal:                                  "'baz\\'s'"
[L:  1, P: 30]      |                [META] dedent:
[L:  1, P: 30]      |    statement_terminator:                                     ';'
[L:  1, P: 31]      |    newline:                                                  '\n'
```
